ogre2/src/Ogre2BoundingBoxCamera.cc

```diff
@@ -0,0 +1,824 @@
+      private: Ogre::MaterialPtr plainOverlayMaterial;
+
+      /// \brief User Data Key to set the label
+      private: std::string labelKey = "label";
+
+      /// \brief Label for background pixels in the ogre Ids map
+      private: uint32_t backgroundLabel {255};
+
+      /// \brief Ogre2 Scene
+      private: Ogre2ScenePtr scene;
+
+      friend class Ogre2BoundingBoxCamera;
+    };
+    }
+  }
+}
+
+/////////////////////////////////////////////////
+BoundingBoxMaterialSwitcher::BoundingBoxMaterialSwitcher(Ogre2ScenePtr _scene)
+{
+  this->scene = _scene;
+
+  // plain material to switch item's material
+  Ogre::ResourcePtr res =
+    Ogre::MaterialManager::getSingleton().load("ign-rendering/plain_color",
+        Ogre::ResourceGroupManager::DEFAULT_RESOURCE_GROUP_NAME);
+
+  this->plainMaterial = res.staticCast<Ogre::Material>();
+  this->plainMaterial->load();
+
+  // plain overlay material
+  this->plainOverlayMaterial =
+      this->plainMaterial->clone("plain_color_overlay");
+  if (!this->plainOverlayMaterial->getTechnique(0) ||
+      !this->plainOverlayMaterial->getTechnique(0)->getPass(0))
+  {
+    ignerr << "Problem creating selection buffer overlay material"
+        << std::endl;
+    return;
+  }
+  Ogre::Pass *overlayPass =
+      this->plainOverlayMaterial->getTechnique(0)->getPass(0);
+  Ogre::HlmsMacroblock macroblock(*overlayPass->getMacroblock());
+  macroblock.mDepthCheck = false;
+  macroblock.mDepthWrite = false;
+  overlayPass->setMacroblock(macroblock);
+}
+
+/////////////////////////////////////////////////
+BoundingBoxMaterialSwitcher::~BoundingBoxMaterialSwitcher()
+{
+}
+
+////////////////////////////////////////////////
+void BoundingBoxMaterialSwitcher::preRenderTargetUpdate(
+    const Ogre::RenderTargetEvent &/*_evt*/)
+{
+  this->datablockMap.clear();
+  auto itor = this->scene->OgreSceneManager()->getMovableObjectIterator(
+      Ogre::ItemFactory::FACTORY_TYPE_NAME);
+
+  while (itor.hasMoreElements())
+  {
+    Ogre::MovableObject *object = itor.peekNext();
+    Ogre::Item *item = static_cast<Ogre::Item *>(object);
+
+    // get visual from ogre item
+    Ogre::Any userAny = item->getUserObjectBindings().getUserAny();
+
+    if (!userAny.isEmpty() && userAny.getType() == typeid(unsigned int))
+    {
+      VisualPtr visual;
+      try
+      {
+        visual = this->scene->VisualById(Ogre::any_cast<unsigned int>(userAny));
+      }
+      catch(Ogre::Exception &e)
+      {
+        ignerr << "Ogre Error:" << e.getFullDescription() << "\n";
+      }
+      Ogre2VisualPtr ogreVisual = std::dynamic_pointer_cast<Ogre2Visual>(
+        visual);
+
+      // get class user data
+      Variant labelAny = ogreVisual->UserData(this->labelKey);
+
+      int label = this->backgroundLabel;
+      try
+      {
+        label = std::get<int>(labelAny);
+      }
+      catch(std::bad_variant_access &e)
+      {
+        // items with no class are considered background
+        label = this->backgroundLabel;
+      }
```

Comment:
There is a problem in reading UserData of the visual, when the visual has no user data of the label key ( "label" ) .. the `std::get<int>(labelKey)` returns 0 and dosn't through the bad_variant_access exception
Any idea about why is that ?